state/src/account.rs

```diff
@@ -0,0 +1,401 @@
+        self.balance = a;
+    }
+
+    pub fn sub_balance(&mut self, x: U256) {
+        let (a, b) = self.balance.overflowing_sub(x);
+        assert_eq!(b, false);
+        self.balance = a;
+    }
+
+    pub fn set_storage(&mut self, key: H256, value: H256) {
+        self.storage_changes.insert(key, value);
+    }
+
+    pub fn get_storage_at_backend<B: DB>(
+        &mut self,
+        db: &mut B,
+        key: &H256,
+    ) -> Result<Option<H256>, Error> {
+        let trie = PatriciaTrie::from(db, RLPNodeCodec::default(), &self.storage_root.0)
+            .or_else(|e| Err(Error::DB(format!("{}", e))))?;
+        if let Some(b) = trie.get(key)? {
+            return Ok(Some(From::from(&b[..])));
+        }
+        Ok(None)
+    }
+
+    pub fn get_storage_at_changes(&self, key: &H256) -> Option<H256> {
+        self.storage_changes.get(key).and_then(|e| Some(*e))
+    }
+
+    pub fn get_storage<B: DB>(&mut self, db: &mut B, key: &H256) -> Result<Option<H256>, Error> {
+        if let Some(value) = self.get_storage_at_changes(key) {
+            return Ok(Some(value));
+        }
+        if let Some(value) = self.get_storage_at_backend(db, key)? {
+            return Ok(Some(value));
+        }
+        Ok(None)
+    }
+
+    pub fn commit_storage<B: DB>(&mut self, db: &mut B) -> Result<(), Error> {
+        let mut trie = if self.storage_root == SHA3_NULL_RLP {
+            PatriciaTrie::new(db, RLPNodeCodec::default())
+        } else {
+            PatriciaTrie::from(db, RLPNodeCodec::default(), &self.storage_root.0)?
+        };
+        for (k, v) in self.storage_changes.drain() {
+            if v.is_zero() {
+                trie.remove(&k)?;
+            } else {
+                trie.insert(&k, &v)?;
+            }
+        }
+        self.storage_root = trie.root()?.into();
+        Ok(())
+    }
+
+    pub fn commit_code<B: DB>(&mut self, db: &mut B) -> Result<(), Error> {
+        match (self.code_state == CodeState::Dirty, self.code.is_empty()) {
+            (true, true) => {
+                self.code_size = 0;
+                self.code_state = CodeState::Clean;
+            }
+            (true, false) => {
+                db.insert(&self.code_hash, &self.code)
+                    .or_else(|e| Err(Error::DB(format!("{}", e))))?;
+                self.code_size = self.code.len();
+                self.code_state = CodeState::Clean;
+            }
+            (false, _) => {}
+        }
+        Ok(())
+    }
+
+    pub fn clone_clean(&self) -> StateObject {
+        StateObject {
+            balance: self.balance,
+            nonce: self.nonce,
+            storage_root: self.storage_root,
+            code: self.code.clone(),
+            code_hash: self.code_hash,
+            code_size: self.code_size,
+            code_state: self.code_state,
+            storage_changes: HashMap::new(),
+        }
+    }
+
+    pub fn clone_dirty(&self) -> StateObject {
+        let mut state_object = self.clone_clean();
```

Comment:
direct `self.clone()`

Reply:
I don't understand. There are no `clone` method at StateObject.

Reply:
So why not add a `#[derive(Clone)]`?

Reply:
Thera are two clones: `clone_dirty()` contains storage_changes, and `clone_clean()` not.